rust-checks/src/bin/no-python-tests.rs

```diff
@@ -11,17 +11,37 @@ fn main() -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
-fn analyze_single_file(log: BufReader<ChildStdout>) -> Result<Report, Box<dyn Error>> {
-    let re = Regex::new(r"Ran 0 tests in \d+.\d+s").unwrap();
+fn analyze_single_file(log: BufReader<ChildStdout>, attr: &Attr) -> Result<Report, Box<dyn Error>> {
+    // Looking for two kinds of messages in the log that indicate no tests were run. The
+    // first is from `setuptoolsCheckPhase`, and looks like
+
+    // running build_ext
+    // ----------------------------------------------------------------------
+    // Ran 0 tests in 0.000s
+    // OK
+    // Finished executing setuptoolsCheckPhase
+
+    // The second is from pytest (either invoked manually or from pytestCheckHook)
+    // and looks like
+    // ============================= test session starts ==============================
+    // platform linux -- Python 3.8.8, pytest-6.2.2, py-1.9.0, pluggy-0.13.1
+    // rootdir: /build/humanize-3.1.0, configfile: tox.ini
+    // collected 0 items
+    // ============================ no tests ran in 0.00s =============================
+    let re = Regex::new(r"(Ran 0 tests|no tests ran\x1b\[0m\x1b\[33m) in \d+.\d+s").unwrap();
+    let ansi_escape = Regex::new(r"\x1b\[0m\x1b\[33m").unwrap();
 
     let report = log
         .lines()
         .filter_map(|line| line.ok())
         .filter_map(|s| re.find(&s).and_then(|m| Some(m.as_str().to_string())))
         .map(|m| NixpkgsHammerMessage {
-            msg: format!("Test runner could not discover any test cases: ‘{}’", m),
+            msg: format!(
+                "Test runner could not discover any test cases: ‘{}’",
+                ansi_escape.replace(&m, "")
+            ),
             name: "no-python-tests",
-            locations: vec![],
+            locations: attr.location.iter().map(|x| x.clone()).collect(),
```

Comment:
One limitation of this is that it doesn't really flag the ideal _line_ within the nix expression for the attr. e.g.

```
nix run -f ~/projects/nixpkgs-hammering -c nixpkgs-hammer -f . python38Packages.humanize -e attribute-orderingWhen evaluating attribute ‘python38Packages.humanize’:
warning: no-python-tests
Test runner could not discover any test cases: ‘no tests ran in 0.00s’
Near /home/mcgibbon/projects/nixpkgs/pkgs/development/python-modules/humanize/default.nix:30:0:
   |
30 |     description = "Python humanize utilities";
   | ^
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
```